### PR TITLE
Improve Testability

### PR DESCRIPTION
**Pull request by asomers**
This PR refactors zabbix-smartmontools to make it more testable.  Now, the core code can be unit tested by mocking `process.checkOutput`.  I've also separated the configurable bits from the main Python file into a dedicated config file.  That will make it easier to use zabbix-smartmontools with a configuration management system.  Finally, I've added a setup.py file for installation.

### CODE REVIEW

**Comment by nobodysu** [comment]
I guess it will not work without, at least, any type timeout: https://github.com/nobodysu/zabbix-smartmontools/commit/c07098be5b9896959e3fcda56543940a1eae62b6
And config does not support comments on the same line. Not essential - the comments could be just removed.
I'm passing out, will get back to it.

File: README.md [hunk 1]
```diff
@@ -47,23 +47,28 @@ yum install zabbix-get   # testing
...
+sudo install -m 644 zabbix-smartmontools.conf /usr/local/etc/zabbix42/
+sudo install -m 644 BSD/zabbix_agentd.conf.d/userparameter_smartctl.conf /usr/local/etc/zabbix42/zabbix_agentd.conf.d/
 ```
 
 #### Windows
 ```cmd
-move smartctl-lld.py C:\zabbix-agent\scripts\
-move sender_wrapper.py C:\zabbix-agent\scripts\
-move userparameter_smartctl.conf C:\zabbix-agent\zabbix_agentd.conf.d\
+python3 setup.py install
```
Comment (nobodysu) [inline README.md]:
```suggestion
python setup.py install
```
Only python.exe on Windows

File: smartctl_lld/__init__.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+        # Add collected data
+        jsonData.append({'{#DDRIVESTATUS}':finalD})   # always populate 'DriveStatus' LLD
+        senderData.extend(getSmart_Out[1])
+        jsonData.extend(getSmart_Out[2])
+    return jsonData, senderData
+
+# Read the config file and return a dict of settings
+def parseConfig(path=None):
+    if sys.platform == 'linux':
```
Comment (nobodysu) [inline smartctl_lld/__init__.py]:
```suggestion
    if sys.platform.startswith('linux')::
```
More compatible.

File: smartctl_lld/__init__.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+        try:
+            zabbixdir = glob.glob('/usr/local/etc/zabbix*')[0]
+        except IndexError:
+            raise 'Error: can\'t find path to Zabbix config directory'
+        agent_conf = '%s/zabbix_agentd.conf' % zabbixdir
+        sender_path = "zabbix_sender"
+        sender_py_path = "%s/scripts/sender_wrapper.py" % zabbixdir
+        if not path:
+            path = "%s/zabbix-smartmontools.conf" % zabbixdir
+    elif os.name == 'nt':
```
Comment (nobodysu) [inline smartctl_lld/__init__.py]:
```suggestion
    elif sys.platform == 'win32':
```
Essential for Windows, don't remember why though.

File: zabbix-smartmontools.conf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+# Path to the smartctl binary
+# ctlPath: smartctl
+# On Windows:
+# ctlPath = C:\Program Files\smartmontools\bin\smartctl.exe
+# On FreeBSD
+# ctlPath = '/usr/local/sbin/smartctl
+
+# path to second send script
+# senderPyPath: /etc/zabbix/scripts/sender_wrapper.py              # Linux
```
Comment (nobodysu) [inline zabbix-smartmontools.conf]: System hint comment is making setting like that to fail. We still need the hint.

**Comment by nobodysu** [comment]
```
$ zabbix_get -k smartctl.discovery[getverb,NON] -s windows.lan
...
"NON" smartctl.info[ConfigStatus] "CONFIGURED"
Traceback (most recent call last):
  File "C:\zabbix-agent\scripts\sender_wrapper.py", line 55, in <module>
    send()
  File "C:\zabbix-agent\scripts\sender_wrapper.py", line 44, in send
    stdin=subprocess.PIPE, universal_newlines=True, close_fds=(not isWindows()))
  File "C:\Program Files\Python37\lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Program Files\Python37\lib\subprocess.py", line 1178, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified
```
`zabbix-smartmontools.conf` is commented out.

**Comment by nobodysu** [comment]
```
debian9# python3 setup.py install
Traceback (most recent call last):
  File "setup.py", line 1, in <module>
    from setuptools import setup, find_packages
ImportError: No module named 'setuptools'
```
I'm strongly against non-essential dependencies. Besides, as far as I understand, a module will make the solution release-dependent. E.g. when python will be updated on Windows the solution cease to work.
I understand it might be beneficial for large installations, but for regular user it adds complexity and more rigid in usability. Maybe make module installation optional?

File: smartctl_lld/__init__.py (hunk 3 above)
Comment (asomers) [reply smartctl_lld/__init__.py]: What value does `sys.platform` have on 64-bit windows?  Is it still `win32`?

File: zabbix-smartmontools.conf (hunk 4 above)
Comment (asomers) [reply zabbix-smartmontools.conf]: I don't understand.  Could you please explain?

File: smartctl_lld/__init__.py (hunk 3 above)
Comment (nobodysu) [reply smartctl_lld/__init__.py]: Still `win32`.

File: zabbix-smartmontools.conf (hunk 4 above)
Comment (nobodysu) [reply zabbix-smartmontools.conf]: This fails:
`senderPyPath: /etc/zabbix/scripts/sender_wrapper.py              # Linux`
This don't:
`senderPyPath: /etc/zabbix/scripts/sender_wrapper.py`

**Comment by asomers** [comment]
> ```
> $ zabbix_get -k smartctl.discovery[getverb,NON] -s windows.lan
> ...
> "NON" smartctl.info[ConfigStatus] "CONFIGURED"
> Traceback (most recent call last):
>   File "C:\zabbix-agent\scripts\sender_wrapper.py", line 55, in <module>
>     send()
>   File "C:\zabbix-agent\scripts\sender_wrapper.py", line 44, in send
>     stdin=subprocess.PIPE, universal_newlines=True, close_fds=(not isWindows()))
>   File "C:\Program Files\Python37\lib\subprocess.py", line 775, in __init__
>     restore_signals, start_new_session)
>   File "C:\Program Files\Python37\lib\subprocess.py", line 1178, in _execute_child
>     startupinfo)
> FileNotFoundError: [WinError 2] The system cannot find the file specified
> ```
> 
> `zabbix-smartmontools.conf` is commented out.

It looks like you're trying to run zabbix-smartmontools using my branch, using the wrong copy of sender_wrapper.py?  The line numbers in the stack trace don't match up with this branch.

**Comment by nobodysu** [comment]
Here's the new version which I mentioned earlier:
https://github.com/nobodysu/zabbix-smartmontools/commit/c07098be5b9896959e3fcda56543940a1eae62b6
Without the new file it won't work at all (wrong type for `timeout`).

**Comment by asomers** [comment]
> Here's the new version which I mentioned earlier:
> [c07098b](https://github.com/nobodysu/zabbix-smartmontools/commit/c07098be5b9896959e3fcda56543940a1eae62b6)
> Without the new file it won't work at all (wrong type for `timeout`).

I already handle converint timeout to an int.  See https://github.com/nobodysu/zabbix-smartmontools/pull/27/files#diff-12ab9df7791f61b30669defc646456baR397 .  So what error do you get when running zabbix_get with the testability branch, and what do you mean by "zabbix-smartmontools.conf is commented out."

**Comment by nobodysu** [comment]
1. I think it's `str` when reading from `zabbix-smartmontools.conf`. Can't tell what error at the moment, but new version is required anyway cause sender_wrapper.py can receive both `str` and `int` as timeout.

2. All lines in `zabbix-smartmontools.conf` is commented out and this error appears.

**Comment by asomers** [comment]
>     1. I think it's `str` when reading from `zabbix-smartmontools.conf`. Can't tell what error at the moment, but new version is required anyway cause sender_wrapper.py can receive both `str` and `int` as timeout.

Notice the difference between `config.getint` and `config.get`.  That's what converts the string value to an integer.  Also, since I turned this project into a python module, sender_wrapper is no longer shared between multiple projects.  So we can be assured that it will also receive the timeout as an int.

>     2. All lines in `zabbix-smartmontools.conf` is commented out and this error appears.

zabbix-smartmontools.conf ought to be fully commented out for most users.  If you're seeing an error, it's probably because one of the default values for Windows is wrong.  But I can't tell which, because you're mixing files from two branches.  Could you please rerun zabbix_get on Windows, using only this branch?

**Comment by nobodysu** [comment]
Alright, I'll test it out once more, give me 24 hours.

Any chance to make the module installation optional? This really bugs me.

**Comment by asomers** [comment]
> Alright, I'll test it out once more, give me 24 hours.
> 
> Any chance to make the module installation optional? This really bugs me.

Module installation is key to testability.  It's how the unittests find the code.  It also allows zabbix-smartmontools to use its own copy of sender_wrapper.py (though I could also accomplish that by simply combining the two files).  What I don't like about setuptools is that it doesn't install everything; the user still has to install a few things manually.  That can probably be fixed.  I'll keep working on it.  Ideally the user would be able to `pip install zabbix-smartmontools` and be ready to go.

Would you accept this PR while I keep working on a better setuptools config, or do you want me to finish it first?

**Comment by nobodysu** [comment]
It could be done in later PR.

**Comment by asomers** [comment]
I've made some progress on this front.  However, the way that setuptools packages modules is incompatible with reexecuting sender_wrapper.py.  So I ask that you accept this PR as-is.  Then I'll submit a patch that eliminates the reexecution, and then submit the rest of the setuptools changes.

File: smartctl_lld/__init__.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        try:
+            zabbixdir = glob.glob('/usr/local/etc/zabbix*')[0]
+        except IndexError:
+            raise 'Error: can\'t find path to Zabbix config directory'
+        agent_conf = '%s/zabbix_agentd.conf' % zabbixdir
+        sender_path = "zabbix_sender"
+        sender_py_path = "%s/scripts/sender_wrapper.py" % zabbixdir
+        if not path:
+            path = "%s/zabbix-smartmontools.conf" % zabbixdir
+    elif os.platform == 'win32':
```
Comment (nobodysu) [inline smartctl_lld/__init__.py]: `elif sys.platform == 'win32':`
Slipped through.

File: smartctl_lld/__init__.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+        agent_conf = 'C:\zabbix_agentd.conf'
+        sender_path = "C:\zabbix-agent\bin\win32\zabbix_sender.exe"
+        sender_py_path = "C:\zabbix-agent\scripts\sender_wrapper.py"
+        if not path:
+            path = "C:\zabbix-agent\zabbix-smartmontools.conf"
```
Comment (nobodysu) [inline smartctl_lld/__init__.py]: Backslash was the culprit. `r` is needed for any windows path string.
E.g. `r'C:\zabbix_agentd.conf'`

**Pull request merged**